File: mit_generator.py
import tkinter as tk


class AppWindow(tk.Tk):
    def __init__(self):
        super().__init__()
        self.title('MIT License Generator')
        self.geometry('800x600')

        # Configure grid for responsive layout
        self.grid_columnconfigure(0, weight=1)
        self.grid_rowconfigure(1, weight=1)

        # Create input section
        input_frame = tk.Frame(self, bg='#f5f5f5', bd=0)
        input_frame.grid(row=0, column=0, padx=20, pady=20, sticky='ew')
        input_frame.grid_columnconfigure(0, weight=1)

        # Input labels and fields
        tk.Label(input_frame, text='Copyright (c)').grid(
            row=0, column=0, sticky='e', padx=5, pady=5)

        self.year_var = tk.StringVar()
        year_entry = tk.Entry(
            input_frame, textvariable=self.year_var, width=10, font=('Arial', 14))
        year_entry.grid(row=0, column=1, sticky='w', padx=5, pady=5)

        self.author_var = tk.StringVar()
        author_entry = tk.Entry(
            input_frame, textvariable=self.author_var, width=40, font=('Arial', 14))
        author_entry.grid(row=0, column=2, sticky='ew', padx=5, pady=5)
        input_frame.grid_columnconfigure(2, weight=1)

        # Buttons frame
        button_frame = tk.Frame(input_frame, bg='#f5f5f5')
        button_frame.grid(row=1, column=0, columnspan=3, pady=10)

        self.generate_btn = tk.Button(button_frame, text='Generate License', command=self.generate_license,
                                      bg='#007bff', fg='white', font=('Arial', 12), relief='flat', state='disabled')
        self.generate_btn.pack(side='left', padx=5)

        self.copy_btn = tk.Button(button_frame, text='Copy to Clipboard', command=self.copy_to_clipboard,
                                  bg='#007bff', fg='white', font=('Arial', 12), relief='flat', state='disabled')
        self.copy_btn.pack(side='left', padx=5)

        # Output section
        output_frame = tk.Frame(self)
        output_frame.grid(row=1, column=0, padx=20,
                          pady=(0, 20), sticky='nsew')
        output_frame.grid_columnconfigure(0, weight=1)
        output_frame.grid_rowconfigure(0, weight=1)

        # License output text area
        self.license_text = tk.Text(output_frame, font=('Courier New', 11), wrap='word',
                                    bg='#f8f9fa', borderwidth=1, relief='solid')
        self.license_text.grid(row=0, column=0, sticky='nsew')
        self.license_text.insert(
            '1.0', 'Please enter the year and author above to generate the MIT License text.')
        self.license_text.config(state='disabled')

        # Scrollbar
        scrollbar = tk.Scrollbar(output_frame, command=self.license_text.yview)
        scrollbar.grid(row=0, column=1, sticky='ns')
        self.license_text.config(yscrollcommand=scrollbar.set)

        # Message label
        self.message_var = tk.StringVar()
        self.message_label = tk.Label(
            self, textvariable=self.message_var, font=('Arial', 12))
        self.message_label.grid(row=2, column=0, padx=20,
                                pady=(0, 10), sticky='w')

        # Event bindings
        self.year_var.trace_add('write', self.validate_inputs)
        self.author_var.trace_add('write', self.validate_inputs)

        year_entry.bind('<KeyPress>', self.validate_year_keypress)

        # Keyboard shortcut for generate (Ctrl+Enter)
        self.bind('<Control-Return>', lambda _: self.generate_license()
                  if self.generate_btn['state'] != 'disabled' else None)

    def validate_inputs(self, name, index, mode):
        import re
        
        _ = name, index, mode
        year = self.year_var.get().strip()
        author = self.author_var.get().strip()

        # Basic validation
        is_valid_year = re.match(r'^\d{4}$', year) is not None
        is_valid_author = len(author) > 0

        if is_valid_year and is_valid_author:
            self.generate_btn.config(state='normal')
        else:
            self.generate_btn.config(state='disabled')

    def validate_year_keypress(self, event):
        # Only allow numbers and backspace
        if not (event.char.isdigit() or event.keysym == 'BackSpace'):
            return 'break'

    def generate_license(self):
        year = self.year_var.get().strip()
        author = self.author_var.get().strip()

        license_text = f"MIT License\n\nCopyright (c) {year} {author}\n\nPermission is hereby granted, free of charge, to any person obtaining a copy\nof this software and associated documentation files (the \"Software\"), to deal\nin the Software without restriction, including without limitation the rights\nto use, copy, modify, merge, publish, distribute, sublicense, and/or sell\ncopies of the Software, and to permit persons to whom the Software is\nfurnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,\nFITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE\nAUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER\nLIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,\nOUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE\nSOFTWARE."

        self.license_text.config(state='normal')
        self.license_text.delete('1.0', tk.END)
        self.license_text.insert('1.0', license_text)
        self.license_text.config(state='disabled')

        self.copy_btn.config(state='normal')
        self.show_message('License generated successfully!', 'success')

    def copy_to_clipboard(self):
        license_content = self.license_text.get('1.0', tk.END).strip()
        if license_content:
            self.clipboard_clear()
            self.clipboard_append(license_content)
            self.update()  # Update the clipboard
            self.show_message('License copied to clipboard!', 'success')
        else:
            self.show_message('No license to copy', 'error')

    def show_message(self, text, message_type):
        if message_type == 'success':
            self.message_label.config(fg='#28a745')
        else:
            self.message_label.config(fg='#dc3545')

        self.message_var.set(text)

        # Clear message after 3 seconds
        self.after(3000, lambda: self.message_var.set(''))


if __name__ == '__main__':
    AppWindow().mainloop()

File: mit_generator.rb
require 'ruby2d'
require 'clipboard'

# Configuration
VP_WIDTH, VP_HEIGHT = 800, 600

# Global state variables
$year = ""
$author = ""
$license_text = "Please enter the year and author above to generate the MIT License text."
$message = ""
$message_type = ""
$message_timer = 0

# UI elements
$year_input = nil
$author_input = nil
$generate_button = nil
$copy_button = nil
$license_output = nil
$message_display = nil

# Input state
$active_input = nil
$cursor_position = 0
$cursor_visible = true
$cursor_timer = 0

# Setup
set(
  title: 'MIT License Generator',
  width: VP_WIDTH,
  height: VP_HEIGHT,
  background: 'white'
)

# Create UI elements

def create_ui
  # Input section background
  Rectangle.new(
    x: 20,
    y: 20,
    width: VP_WIDTH - 40,
    height: 120,
    color: '#f5f5f5'
  )
  
  # Copyright label
  Text.new(
    'Copyright (c)',
    x: 30,
    y: 60,
    font: Font.default,
    size: 18,
    color: 'black'
  )
  
  # Year input field
  $year_field = Rectangle.new(
    x: 180,
    y: 50,
    width: 100,
    height: 30,
    color: 'white'
  )
  
  # Year input text
  $year_input = Text.new(
    $year,
    x: 185,
    y: 53,
    size: 16,
    color: 'black'
  )
  
  # Author input field
  $author_field = Rectangle.new(
    x: 300,
    y: 50,
    width: VP_WIDTH - 340,
    height: 30,
    color: 'white'
  )
  
  # Author input text
  $author_input = Text.new(
    $author,
    x: 305,
    y: 53,
    size: 16,
    color: 'black'
  )
  
  # Generate button
  $generate_button = Rectangle.new(
    x: 30,
    y: 100,
    width: 150,
    height: 35,
    color: '#6c757d',  # Disabled by default
  )
  
  Text.new(
    'Generate License',
    x: $generate_button.x + 20,
    y: $generate_button.y + 7,
    size: 16,
    color: 'white'
  )
  
  # Copy button
  $copy_button = Rectangle.new(
    x: 200,
    y: 100,
    width: 150,
    height: 35,
    color: '#6c757d',  # Disabled by defaul
  )
  
  Text.new(
    'Copy to Clipboard',
    x: $copy_button.x + 15,
    y: $copy_button.y + 7,
    size: 16,
    color: 'white'
  )
  
  # License output area
  Rectangle.new(
    x: 20,
    y: 160,
    width: VP_WIDTH - 40,
    height: VP_HEIGHT - 200,
    color: '#f8f9fa'
  )
  
  # License text - this will be updated dynamically
  update_license_output
  
  # Message area
  $message_display = Text.new(
    $message,
    x: 30,
    y: VP_HEIGHT - 30,
    size: 16,
    color: 'black'
  )
end

# Update license output text
def update_license_output
  $license_output.remove if $license_output
  
  # Draw license text directly
  $license_output = Text.new(
    $license_text,
    x: 40,
    y: 180,
    size: 14,
    color: 'black'
  )
end

# Update message display
def update_message
  $message_display.remove if $message_display
  
  $message_display = Text.new(
    $message,
    x: 30,
    y: VP_HEIGHT - 30,
    size: 16,
    color: $message_type == 'success' ? '#28a745' : '#dc3545'
  )
end

# Validate inputs
def validate_inputs
  valid_year = $year.match?(/^\d{4}$/)
  valid_author = !$author.strip.empty?
  
  if valid_year && valid_author
    $generate_button.color = '#007bff'
  else
    $generate_button.color = '#6c757d'
  end
  
  valid_year && valid_author
end

# Generate license text
def generate_license
  if validate_inputs
    $license_text = <<~LICENSE
MIT License

Copyright (c) #{$year} #{$author}

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
LICENSE
    
    update_license_output
    $copy_button.color = '#007bff'
    show_message("License generated successfully!", "success")
  end
end

# Copy license to clipboard
def copy_to_clipboard
  begin
    Clipboard.copy($license_text)
    show_message("License copied to clipboard!", "success")
  rescue => e
    show_message("Failed to copy license to clipboard", "error")
  end
end

# Show message
def show_message(text, type)
  $message = text
  $message_type = type
  $message_timer = 300  # 3 seconds at 60fps
  update_message
end

# Update message timer
def update_message_timer
  if $message_timer > 0
    $message_timer -= 1
  elsif $message_timer == 0
    $message = ""
    update_message
    $message_timer = -1
  end
end

# Update cursor timer
def update_cursor
  $cursor_timer += 1
  if $cursor_timer >= 30  # Blink every 0.5 seconds at 60fps
    $cursor_visible = !$cursor_visible
    $cursor_timer = 0
  end
end

# Handle mouse click
def handle_mouse_click(x, y)
  # Check year input field
  if x >= 180 && x <= 280 && y >= 50 && y <= 80
    $active_input = :year
    $cursor_position = $year.length
  # Check author input field
  elsif x >= 300 && x <= VP_WIDTH - 40 && y >= 50 && y <= 80
    $active_input = :author
    $cursor_position = $author.length
  # Check generate button
  elsif x >= $generate_button.x && x <= $generate_button.x + $generate_button.width &&
        y >= $generate_button.y && y <= $generate_button.y + $generate_button.height
    $active_input = nil
    generate_license
  # Check copy button
  elsif x >= $copy_button.x && x <= $copy_button.x + $copy_button.width &&
        y >= $copy_button.y && y <= $copy_button.y + $copy_button.height
    $active_input = nil
    copy_to_clipboard
  else
    $active_input = nil
  end
  
  # Redraw input fields with active state
  draw_input_fields
end

# Handle key press
def handle_key_press(key)
  return unless $active_input
  
  case key
  when 'backspace'
    if $active_input == :year
      $year = $year.chop if $year.length > 0
      $cursor_position = [$cursor_position - 1, 0].max
    else
      $author = $author.chop if $author.length > 0
      $cursor_position = [$cursor_position - 1, 0].max
    end
  when 'left'
    $cursor_position = [$cursor_position - 1, 0].max
  when 'right'
    max_pos = $active_input == :year ? $year.length : $author.length
    $cursor_position = [$cursor_position + 1, max_pos].min
  when 'return'
    # Generate license if both inputs are valid
    generate_license if validate_inputs
  else
    # Only allow digits for year input
    if $active_input == :year
      return unless key =~ /\d/ && $year.length < 4
      $year += key
      $cursor_position += 1
    elsif key.length == 1 && $author.length < 100
      # Allow any character for author input
      $author += key
      $cursor_position += 1
    end
  end
  
  # Redraw input fields and validate
  draw_input_fields
  validate_inputs
end

# Draw input fields
def draw_input_fields
  # Remove existing input text
  $year_input.remove if $year_input
  $author_input.remove if $author_input
  
  # Redraw year input
  $year_input = Text.new(
    $year,
    x: 185,
    y: 53,
    size: 16,
    color: 'black'
  )
  
  # Redraw author input
  $author_input = Text.new(
    $author,
    x: 305,
    y: 53,
    size: 16,
    color: 'black'
  )
  
  # Active input indication removed due to Ruby2D limitations
  $year_field.color = $active_input == :year ? '#98fb98' : 'white'
  $author_field.color = $active_input == :author ? '#98fb98' : 'white'
end

# Mouse click handler
on :mouse_down do |event|
  handle_mouse_click(event.x, event.y)
end

# Key press handler
on :key_down do |event|
  handle_key_press(event.key)
end

# Update handler (called every frame)
update do
  update_message_timer
  update_cursor
  
  # Redraw input fields to update cursor
  draw_input_fields if $active_input
end

# Create the UI and start the application
create_ui
show